contrib/src/main/java/com/datatorrent/contrib/cassandra/AbstractUpsertOutputOperator.java

```diff
@@ -0,0 +1,1401 @@
+
+  transient Map<DataType.Name, List<String>> dataTypeGroupedColumns;
+
+  transient Set<String> pkColumnNames;
+
+  transient Set<String> counterColumns;
+
+  transient Set<String> collectionColumns;
+
+  transient Set<String> listColumns;
+
+  transient Set<String> mapColumns;
+
+  transient Set<String> setColumns;
+
+  transient Set<String> userDefinedTypeColumns;
+
+  transient Set<String> regularColumns;
+
+  transient Map<String, String> pojoFieldNamesToCassandraColumns;
+
+  Map<String, Object> getters;
+
+  Map<String, TypeCodec> codecsForCassandraColumnNames;
+
+  transient Map<Long, PreparedStatement> preparedStatementTypes;
+  transient Class<?> tuplePayloadClass;
+
+  private transient Logger LOG = LoggerFactory.getLogger(AbstractUpsertOutputOperator.class);
+
+  // Metrics
+
+  @AutoMetric
+  transient long ignoredRequestsDuetoIfExistsCheck = 0;
+
+  @AutoMetric
+  transient long successfullWrites = 0;
+
+  @AutoMetric
+  long totalWriteRequestsAttempted = 0;
+
+  @AutoMetric
+  transient int numberOfHostsWrittenTo = 0;
+
+  @AutoMetric
+  transient double fifteenMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double fiveMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double oneMinuteWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient double meanWriteRateLatency = 0.0;
+
+  @AutoMetric
+  transient long totalIgnoresInThisWindow = 0;
+
+  @AutoMetric
+  long totalIgnoresSinceStart = 0;
+
+  @AutoMetric
+  transient long totalWriteTimeoutsInThisWindow = 0;
+
+  @AutoMetric
+  long totalWriteTimeoutsSinceStart = 0;
+
+  @AutoMetric
+  transient long totalWriteRetriesInThisWindow =  0;
+
+  @AutoMetric
+  long totalWriteRetriesSinceStart = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyOne = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyTwo = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyThree = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyAll = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyLocalOne = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyQuorum = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyLocalQuorum = 0;
+
+  @AutoMetric
+  transient long writeWithConsistencyLocalSerial = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyEachQuorum = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencySerial = 0;
+
+  @AutoMetric
+  transient long writesWithConsistencyAny = 0;
+
+  transient Set<Host> uniqueHostsWrittenToInCurrentWindow;
+
+
+  @InputPortFieldAnnotation
+  public final transient DefaultInputPort<UpsertExecutionContext> input = new DefaultInputPort<UpsertExecutionContext>()
+  {
+    @Override
+    public void process(final UpsertExecutionContext tuple)
+    {
+      if (!isEligbleForPassivation(tuple)) {
+        return;
+      }
+      BoundStatement stmnt = setDefaultsAndPrepareBoundStatement(tuple);
+      ResultSet result = session.execute(stmnt);
```

Comment:
Yes. It has been purposefully designed to update one row at a time. The main reason is to support the following use cases:
1. This operator is letting the end users specify an operational context and not just setting the values. For example TTL, consistency etc, handle nulls etc 
2. This can be done at the singular level. Setting this at a batch level does not look to be a simplistic view of the implementation. 
3. Also some aspects like TTL cannot be set at a record level if it is a batch statement. This means the feature of the operator needs to be compromised. Since cassandra TTL is at a cell level and a given row can have varying TTLs across its columns, I believe giving a per row flexibility would be more beneficial for the end user. 
3. Regarding the question on performance, sure there is a little overhead ( for the network call from the operator JVM instance node to the Cassandra cluster coordinator node ) but it would be interesting to see what are the overheads with respect to a batch ( Because all statements in a batch become one transaction , it looks like the coordinator node does a lot more tricks to manage its transaction state like a multi-phase handshake across all participating cassandra nodes for a given batch statement) 
5. Perhaps prepared statements mitigate this to a good extent. 

I was able to reach 5000 QPS on a 10 node cassandra cluster with 10 instances of Cassandra operator ( different implementation but taking the same approach of a single statement as opposed to a batch ) and all this without any stress on both the cassandra cluster and the Apex containers. ( I understand that the numbers here are a bit non-descriptive with respect to the hardware used etc but just wanted to point out that single statement executions are not that bad perhaps ) .